cmd/uplink/cmd/root.go

```diff
@@ -37,32 +37,83 @@ func addCmd(cmd *cobra.Command, root *cobra.Command) *cobra.Command {
 	root.AddCommand(cmd)
 
 	defaultConfDir := fpath.ApplicationDir("storj", "uplink")
-	defaultIdentityDir := fpath.ApplicationDir("storj", "identity", "uplink")
 
 	confDirParam := cfgstruct.FindConfigDirParam()
 	if confDirParam != "" {
 		defaultConfDir = confDirParam
 	}
-	identityDirParam := cfgstruct.FindIdentityDirParam()
-	if identityDirParam != "" {
-		defaultIdentityDir = identityDirParam
-	}
 
-	cfgstruct.Bind(cmd.Flags(), &cfg, defaults, cfgstruct.ConfDir(defaultConfDir), cfgstruct.IdentityDir(defaultIdentityDir))
+	cfgstruct.Bind(cmd.Flags(), &cfg, defaults, cfgstruct.ConfDir(defaultConfDir))
+
 	return cmd
 }
 
-// Metainfo loads the storj.Metainfo
-//
-// Temporarily it also returns an instance of streams.Store until we improve
-// the metainfo and streas implementations.
-func (c *UplinkFlags) Metainfo(ctx context.Context) (storj.Metainfo, streams.Store, error) {
+// NewUplink returns a pointer to a new Client with a Config and Uplink pointer on it and an error.
+func (c *UplinkFlags) NewUplink(ctx context.Context, config *libuplink.Config) (*libuplink.Uplink, error) {
+	return libuplink.NewUplink(ctx, config)
+}
+
+// GetProject returns a *libuplink.Project for interacting with a specific project
+func (c *UplinkFlags) GetProject(ctx context.Context) (*libuplink.Project, error) {
+	apiKey, err := libuplink.ParseAPIKey(c.Client.APIKey)
+	if err != nil {
+		return nil, err
+	}
+
+	satelliteAddr := c.Client.SatelliteAddr
+
 	identity, err := c.Identity.Load()
+	if err != nil {
+		return nil, err
+	}
+
+	identityVersion, err := identity.Version()
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := &libuplink.Config{}
+
+	cfg.Volatile.TLS = struct {
+		SkipPeerCAWhitelist bool
+		PeerCAWhitelistPath string
+	}{
+		SkipPeerCAWhitelist: !c.TLS.UsePeerCAWhitelist,
+		PeerCAWhitelistPath: c.TLS.PeerCAWhitelistPath,
+	}
+
+	cfg.Volatile.UseIdentity = identity
+	cfg.Volatile.IdentityVersion = identityVersion
+	cfg.Volatile.MaxInlineSize = c.Client.MaxInlineSize
+	cfg.Volatile.MaxMemory = c.RS.MaxBufferMem
+
+	uplink, err := c.NewUplink(ctx, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := &libuplink.ProjectOptions{}
+
+	encKey := new(storj.Key)
+	copy(encKey[:], c.Enc.Key)
+	opts.Volatile.EncryptionKey = encKey
+
+	return uplink.OpenProject(ctx, satelliteAddr, apiKey, opts)
+}
+
+// GetProjectAndBucket returns a *libuplink.Bucket for interacting with a specific project's bucket
+func (c *UplinkFlags) GetProjectAndBucket(ctx context.Context, bucketName string, access libuplink.EncryptionAccess) (*libuplink.Project, *libuplink.Bucket, error) {
+	project, err := c.GetProject(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	bucket, err := project.OpenBucket(ctx, bucketName, &access)
 	if err != nil {
 		return nil, nil, err
```

Comment:
if this fails you probably want to close the project before returning an error. i'm assuming the return after is delegating project/bucket closing to the caller

Reply:
When we implement close on the bucket and project will they allow closing twice without error?

Reply:
oh i see, you have a defer that closes the project in the non-nil error case. okay, :+1: